Sources/Signal.swift

```diff
@@ -89,18 +98,32 @@ public final class Signal<Value, Error: Swift.Error> {
 					// whichever sender loses the battle of acquring `sendLock` is
 					// guaranteed to see a status of `terminated`, and eventually be
 					// blocked.
-					signal.status = .terminating(SignalTerminationSnapshot(state: state, event: event))
+					signal.status = .terminating(SignalState(observers: state.observers,
+					                                         retaining: state.retaining,
+					                                         associatedStorage: event))
+					signal.statusLock.unlock()
 
 					if signal.sendLock.try() {
-						if case let .terminating(state) = signal.status {
-							state.send()
-							signal.status = .terminated
+						// Is it terminating? (False positive is allowed.)
+						if case .terminating = signal.status {
+							// Acquire `statusLock`, and handle the termination if the signal
+							// is still at `terminating` state.
+							signal.statusLock.lock()
+							if case let .terminating(state) = signal.status {
```

Comment:
As noted in the comment, read-copy-update is used to synchronise `status`. Here we take advantage of the wait-free reading to `status`, so that we acquire the `statusLock` only if `terminating` is being seen by the sender.

The second confirmation after acquiring `statusLock` is to ensure that the termination is really not yet handled, since it could have observed `terminating` while some other concurrent sender is handling it.